python/dgl/optim/pytorch/sparse_optim.py

```diff
@@ -474,6 +474,40 @@ def zero_grad(self):
         """clean grad cache"""
         self._clean_grad = True
 
+    def state_dict(self, **kwargs):  # pylint: disable=unused-argument
+        """Return a copy of the whole optimizer states stored in CPU memory.
+        If this is a multi-processing instance, the states will be returned in
+        shared memory. If the underlying embedding is currently stored on
+        multiple GPUs, all processes must call this method in the same order.
+
+        NOTE: This method must be called by all processes sharing the
+        underlying embedding, or it may result in a deadlock.
+
+        Returns
+        -------
+        dictionary of optimizer states
+            The optimizer states stored in CPU memory.
+        """
+        return {emb.name: emb._all_get_optm_state() for emb in self._params}
```

Comment:
Do we need also store basic hyper parameters like `lr`, `beta`?

Reply:
Updated.

Reply:
Could we make the format uniformed with [dist embedding state dict](https://github.com/dmlc/dgl/blob/master/python/dgl/distributed/optim/pytorch/sparse_optim.py#L66-L82)? So that user won't be confused to get  two different formats.

Reply:
Some thoughts on this:
1. The current format follows PyTorch optimizer's style (keys are "state" and "param_groups"):
```
{
    "state": {
        emb_name: state_tensor,
        ...
    },
    "param_groups": {
        "lr": ,
        ...
    }
}
```
2. Different from the **local** state dict of dist embedding, the state dict here doesn't have "ids", so I feel the following structure is a bit complicated for it:
```
{
    "params": {
        ...,
    },
    "emb_states": {
        emb_name: {
            "ids": ...,  # don't need it 
            "state": {   # the example in your link misses this key
                state_name: state_tensor, 
                ...,
            },
        },
    }
}
```

Reply:
I see, follow pytorch style is a good choice.